File: ios/DisplaySessionManager.swift
import Foundation
import MWDATCore
import MWDATDisplay

public typealias EventEmitterDisplay = (String, [String: Any]) -> Void

@MainActor
public final class DisplaySessionManager {
    public static let shared = DisplaySessionManager()

    private let logger = EMWDATLogger.shared
    private var displays: [String: Display] = [:]
    private var stateTokens: [String: AnyListenerToken] = [:]
    private var eventEmitter: EventEmitterDisplay?

    private init() {}

    public func setEventEmitter(_ emitter: @escaping EventEmitterDisplay) {
        self.eventEmitter = emitter
    }

    // MARK: - Display Capability Control

    public func addDisplayToSession(sessionId: String) async throws {
        guard let session = WearablesManager.shared.getSession(sessionId: sessionId) else {
            throw DisplaySessionManagerError.sessionNotFound(sessionId)
        }

        let display = try session.addDisplay()
        displays[sessionId] = display

        stateTokens[sessionId] = display.statePublisher.listen { [weak self] state in
            Task { @MainActor in
                self?.handleStateChange(sessionId: sessionId, state: state)
            }
        }

        await display.start()

        logger.info("Display", "Display added and started", context: ["sessionId": sessionId])
    }

    public func removeDisplayFromSession(sessionId: String) async {
        await displays[sessionId]?.stop()
        destroyDisplay(sessionId: sessionId)
        emitEvent("onDisplayStateChange", ["sessionId": sessionId, "state": "stopped"])
        logger.info("Display", "Display removed from session", context: ["sessionId": sessionId])
    }

    public func sendDisplayContent(sessionId: String, contentTree: [String: Any]) async throws {
        guard let display = displays[sessionId] else {
            throw DisplaySessionManagerError.displayNotFound(sessionId)
        }

        guard let rootView = DisplayContentBuilder.build(from: contentTree, onInteraction: { [weak self] interactionId in
            Task { @MainActor in
                self?.emitEvent("onDisplayInteraction", [
                    "sessionId": sessionId,
                    "interactionId": interactionId
                ])
            }
        }) as? FlexBox else {
            throw DisplaySessionManagerError.invalidContentTree("Root node must be a flexBox")
        }

        try await display.send(rootView)
    }

    public func destroy() {
        for sessionId in Array(displays.keys) {
            destroyDisplay(sessionId: sessionId)
        }
    }

    // MARK: - Private

    private func handleStateChange(sessionId: String, state: DisplayState) {
        let mapped = mapDisplayState(state)
        logger.info("Display", "State changed", context: ["sessionId": sessionId, "state": mapped])
        emitEvent("onDisplayStateChange", ["sessionId": sessionId, "state": mapped])

        if state == .stopped || state == .stopping {
            // Only fully destroy on stopped, not stopping
        }
        if state == .stopped {
            destroyDisplay(sessionId: sessionId)
        }
    }

    private func destroyDisplay(sessionId: String) {
        stateTokens[sessionId] = nil
        displays[sessionId] = nil
    }

    private func mapDisplayState(_ state: DisplayState) -> String {
        switch state {
        case .stopped: return "stopped"
        case .starting: return "starting"
        case .started: return "started"
        case .stopping: return "stopping"
        @unknown default: return "stopped"
        }
    }

    private func emitEvent(_ name: String, _ body: [String: Any]) {
        eventEmitter?(name, body)
    }
}

// MARK: - Errors

public enum DisplaySessionManagerError: LocalizedError {
    case sessionNotFound(String)
    case displayNotFound(String)
    case invalidContentTree(String)

    public var errorDescription: String? {
        switch self {
        case .sessionNotFound(let id): return "Session not found: \(id)"
        case .displayNotFound(let id): return "No active display for session: \(id)"
        case .invalidContentTree(let reason): return "Invalid content tree: \(reason)"
        }
    }
}

File: android/src/main/java/expo/modules/emwdat/DisplaySessionManager.kt
package expo.modules.emwdat

import com.meta.wearable.dat.display.Display
import com.meta.wearable.dat.display.DisplayConfiguration
import com.meta.wearable.dat.display.types.DisplayError
import com.meta.wearable.dat.display.types.DisplayState
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch

object DisplaySessionManager {
    private val logger = EMWDATLogger

    private val displays: MutableMap<String, Display> = mutableMapOf()
    private val stateJobs: MutableMap<String, Job> = mutableMapOf()

    private var scope: CoroutineScope? = null
    private var eventEmitter: EventEmitter? = null

    fun setEventEmitter(emitter: EventEmitter) {
        eventEmitter = emitter
    }

    fun setScope(s: CoroutineScope) {
        scope = s
    }

    fun addDisplayToSession(sessionId: String) {
        val session = WearablesManager.getSession(sessionId)
            ?: throw IllegalArgumentException("Session not found: $sessionId")

        var display: Display? = null
        session.addDisplay(DisplayConfiguration()).fold(
            onSuccess = { d -> display = d },
            onFailure = { error -> throw Exception("Failed to add display: $error") }
        )
        val activeDisplay = display ?: throw Exception("Failed to add display to session")
        displays[sessionId] = activeDisplay

        val currentScope = scope ?: throw IllegalStateException("Module scope not available")
        stateJobs[sessionId] = currentScope.launch {
            activeDisplay.state.collect { state ->
                handleStateChange(sessionId, state)
            }
        }

        logger.info("Display", "Display added to session", mapOf("sessionId" to sessionId))
    }

    fun removeDisplayFromSession(sessionId: String) {
        WearablesManager.getSession(sessionId)?.removeDisplay()
        destroyDisplay(sessionId)
        emitEvent("onDisplayStateChange", mapOf("sessionId" to sessionId, "state" to "stopped"))
        logger.info("Display", "Display removed from session", mapOf("sessionId" to sessionId))
    }

    suspend fun sendDisplayContent(sessionId: String, contentTree: Map<String, Any>) {
        val display = displays[sessionId]
            ?: throw IllegalArgumentException("No active display for session: $sessionId")

        val result = display.sendContent {
            DisplayContentBuilder.build(this, contentTree) { interactionId ->
                emitEvent(
                    "onDisplayInteraction",
                    mapOf("sessionId" to sessionId, "interactionId" to interactionId)
                )
            }
        }

        result.fold(
            onSuccess = { /* sent */ },
            onFailure = { error ->
                val errorCode = mapDisplayError(error)
                emitEvent("onDisplayError", mapOf("sessionId" to sessionId, "error" to errorCode))
                throw Exception("sendDisplayContent failed: $error")
            }
        )
    }

    private fun handleStateChange(sessionId: String, state: DisplayState) {
        val mapped = mapDisplayState(state)
        logger.info("Display", "State changed", mapOf("sessionId" to sessionId, "state" to mapped))
        emitEvent("onDisplayStateChange", mapOf("sessionId" to sessionId, "state" to mapped))

        if (state == DisplayState.CLOSED || state == DisplayState.STOPPED) {
            destroyDisplay(sessionId)
        }
    }

    private fun destroyDisplay(sessionId: String) {
        stateJobs[sessionId]?.cancel()
        stateJobs.remove(sessionId)
        displays.remove(sessionId)
    }

    fun destroy() {
        for (sessionId in displays.keys.toList()) {
            destroyDisplay(sessionId)
        }
    }

    private fun mapDisplayState(state: DisplayState): String = when (state) {
        DisplayState.STOPPED -> "stopped"
        DisplayState.STARTING -> "starting"
        DisplayState.STARTED -> "started"
        DisplayState.STOPPING -> "stopping"
        DisplayState.CLOSED -> "stopped"
        else -> "stopped"
    }

    private fun mapDisplayError(error: DisplayError): String = when (error) {
        DisplayError.CAPABILITY_DENIED -> "capabilityDenied"
        DisplayError.DEVICE_DISCONNECTED -> "deviceDisconnected"
        DisplayError.INVALID_SESSION_STATE -> "invalidSessionState"
        DisplayError.RENDERING_FAILED -> "renderingFailed"
        else -> "unexpectedError"
    }

    private fun emitEvent(name: String, body: Map<String, Any>) {
        eventEmitter?.invoke(name, body)
    }
}
